### PR TITLE
Refactor dashboard and settings tabs into routes

### PR DESCRIPTION
**Pull request by somebody1234**
### Pull Request Description
- Close https://github.com/enso-org/cloud-v2/issues/1410
  - Enables links in the future to link directory to e.g. a specific settings page
    - Deep links *do not* support these new links yet, however it would be simple enough to add support
  - Allow "settings" icon in Category Switcher to do `navigate('/settings/local')` rather than requiring a workaround
  - Remove obsolete `searchParamsState` hooks

### Important Notes
- When navigating, the URL search parameters are lost (and so they are missing by the time GUI reads them).
  - GUI team should review whether this is an issue for the time being
  - We should probably rethink how CLI parameters should be passed to GUI (if at all)

### Testing instructions
Apart from testing all pages (all settings pages, all drive categories, editor tab):
- Make sure you stay on the same page (including settings subpages and drive categories) after a reload
- Test opening and closing projects and switching tabs with projects open
- Make sure that projects are still reopened after reload (and do not magically appear after reload if closed
- Make sure settings tab shows as selected after reload (because the active tab is done by adding a clip mask to the darker background behind it, so if there is a race condition then it may get out of sync)

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [x] ~~The documentation has been updated, if necessary.~~
- [x] ~~Screenshots/screencasts have been attached, if there are any visual changes. For interactive or animated visual changes, a screencast is preferred.~~
- [x] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      [TypeScript](https://github.com/enso-org/enso/blob/develop/docs/style-guide/typescript.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides. In case you are using a language not listed above, follow the [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guide.
- [x] ~~Unit tests have been written where possible.~~

### CODE REVIEW

**Comment by MrFlashAccount** [comment]
> Remove obsolete searchParamsState hooks

Disagree it's obsolete. It's a very simple state manager on top of the query parameters.

**Comment by MrFlashAccount** [comment]
Also, I'm not sure that this PR is worth time spending on: 

1. React router is about to be released (I assume it will be public in the next 2 months)
2. Tanstack router has it out of the box
3. When we'll start the transition from the react-router to tanstack, we'll probably remove this code completely

**Comment by somebody1234** [comment]
> It's a very simple state manager on top of the query parameters.

yes, but this PR replaces query params with routes (query params are lost when navigating away so i'm not sure they're useful anymore)

> tanstack router is about to be released 

the point of this PR is to refactor the URLs in the app, the idea is the same regardless of the underlying framework

> When we'll start the transition from the react-router to tanstack, we'll probably remove this code completely

again no, we will need URLs in some form, see the point above

**Comment by MrFlashAccount** [comment]
> yes, but this PR replaces query params with routes (query params are lost when navigating away so i'm not sure they're useful anymore)

It's okay, it's the lifecycle of this kind of state. It has its own advantages and disadvantages, e.g. it's preserved in history, restores on navigating back & forth, you can set this state outside (by navigating using these parameters), and users can share this state with others. None of the existing stores have these abilities.

**Comment by somebody1234** [comment]
fair enough

**Comment by MrFlashAccount** [comment]
> the point of this PR is to refactor the URLs in the app, the idea is the same regardless of the underlying framework

So let's then put settings under the `/pages/` folder?

And it's not the same. In this particular case, the route structure will be drastically different compared to the current solution:

1. We'll use nested routes to represent sections, see: https://tanstack.com/router/latest/docs/framework/react/guide/route-trees
2. We'll inline paths into the route configuration and colocate config parts with the according routes, thus appUtis might be even removed.
3. `React aria components` lib supports tanstack router out of the box: https://react-spectrum.adobe.com/react-aria/routing.html#tanstack-router
4. Since I joined enso we haven't faced issues with the current navigation, which means that this part changes less often than the other(thus requiring less immediate effort to spend) -> typing won't make things much better as well and won't solve anything.
5. The main purpose of migration to tanstack router is: easier to work with (react-router API is unobvious and hard to debug), support for context in loaders(we'll use it to prefetch queries), ease of adding guards and paywalls on top of the tanstack API.

So the tanstack solution will be drastically different from the current one and that's why I'm saying that it's not a good idea to invest our time into improving the ergonomics of the current API.


<blockquote><div><strong><a href="https://tanstack.com/router/latest/docs/framework/react/guide/route-trees">Route Trees & Nesting | TanStack Router React Docs</a></strong></div><div>
Like most other routers, TanStack Router uses a nested route tree to match up the URL with the correct component tree to render.

To build a route tree, TanStack Router supports both:</div></blockquote>
<blockquote><div><strong><a href="https://react-spectrum.adobe.com/react-aria/routing.html">Client Side Routing</a></strong></div><div>Documentation for Client Side Routing in the React Aria package.</div></blockquote>

**Comment by somebody1234** [comment]
> So let's then put settings under the `/pages/` folder?

good point, i guess `Drive` and `Editor` belong there too

> We'll use nested routes to represent sections

react router can do that as well, no? so if you have issues with the current structure then that can probably be changed even before the move to tanstack router

> appUtils might be even removed

that is not a problem, `appUtils` is simply a means to an end

> typing won't make things much better

correct, it exists mostly for autocomplete. i'm just not a fan of having to import the paths everywhere.
and again, i cannot stress this enough, but the main point of this PR is to actually use paths for our pages, the other stuff like the type refactors are secondary. if you have issues with the added types i can revert it to the old list of constants

> it's not a good idea to invest our time

except the time has already been spent

**Comment by MrFlashAccount** [comment]
> good point, i guess Drive and Editor belong there too

Definetely

**Comment by MrFlashAccount** [comment]
> react router can do that as well, no? so if you have issues with the current structure then that can probably be changed even before the move to tanstack router

Yes, and I had a few attempts to migrate, but I concluded that it'd be better to spend that effort on migration to tanstack router.

**Comment by somebody1234** [comment]
tl;dr: the main point of this PR is the paths (feedback on the path structure is welcome of course)

this also means the rest of the refactors:
a) are only done because i just prefer things that way over the current structure. i *could* have left things as they were, but i just like that now i can just `navigate()` and the path will be autocompleted for me
b) i'm happy to change if you have better alternatives

**Comment by MrFlashAccount** [comment]
> except the time has already been spent

Yes, and you left me no other choice rather to approve that. I just want to share my thoughts and the way I think about changes like these, the priorities, and the other stuff. Just want to be with you on the same page.

So my point is that next time let's spend our spare time on migration to tanstack instead of trying to improve the current solution

**Comment by somebody1234** [comment]
> no other choice rather to approve that

i don't think it's a big loss to lose less than a day of work fwiw, and again i have to point out that the main change is switching from `http://localhost:8080/?cloud-ide_page=%22settings%22&cloud-ide_SettingsTab=%22organization%22` to `http://localhost:8080/settings/organization` so that the settings page can be navigated to without weird workarounds

> trying to improve the current solution

(again see above, the point of this PR is not to improve DX)

tl;dr:
- the main point of this PR is to change the paths. all other refactors don't matter, they can be changed or reverted easily enough
- this PR represents an *option* that we have. i guess i just assumed that this would be a sensible solution to the workaround that is currently in develop to navigate to `/settings/local`

File: app/dashboard/src/layouts/AssetsTable.tsx [hunk 1]
```diff
@@ -295,11 +295,11 @@ interface DragSelectionInfo {
 // === Category to filter by ===
 // =============================
 
-const CATEGORY_TO_FILTER_BY: Readonly<Record<Category, backendModule.FilterBy | null>> = {
-  [Category.cloud]: backendModule.FilterBy.active,
-  [Category.local]: backendModule.FilterBy.active,
-  [Category.recent]: null,
-  [Category.trash]: backendModule.FilterBy.trashed,
+const CATEGORY_TO_FILTER_BY: Readonly<Record<DriveCategory, backendModule.FilterBy | null>> = {
+  ['cloud']: backendModule.FilterBy.active,
```
Comment (MrFlashAccount) [inline app/dashboard/src/layouts/AssetsTable.tsx]: I think `[]` is not needed anymore

File: app/dashboard/src/pages/dashboard/Dashboard.tsx [hunk 2]
```diff
@@ -150,9 +133,23 @@ function DashboardInner(props: DashboardProps) {
...
+    // This *may* look cleaner as separate routes in the Router, however it would cause the
+    // Dashboard to be unmounted and then remounted, which is not preferred.
+    if (pathname === '/' || pathname === '/drive') {
+      const newCategory =
+        localStorage.get('driveCategory') ??
+        (() => {
+          const shouldDefaultToCloud = user.isEnabled || localBackend == null
+          return shouldDefaultToCloud ? 'cloud' : 'local'
+        })()
+      navigate(`/drive/${newCategory}`)
```
Comment (MrFlashAccount) [inline app/dashboard/src/pages/dashboard/Dashboard.tsx]: I think here should be replace: true?

File: app/dashboard/src/pages/dashboard/Settings/MembersTable.tsx [hunk 3]
```diff
@@ -10,7 +10,7 @@ import * as toastAndLogHooks from '#/hooks/toastAndLogHooks'
 import * as authProvider from '#/providers/AuthProvider'
 import * as textProvider from '#/providers/TextProvider'
 
-import UserRow from '#/layouts/Settings/UserRow'
+import UserRow from '#/pages/dashboard/Settings/UserRow'
```
Comment (MrFlashAccount) [inline app/dashboard/src/pages/dashboard/Settings/MembersTable.tsx]: Can we please use relative imports for internal components? 

What an internal component is:

we have a SettingsPage. Everything under this page uses relative imports when imports siblings/nested components. If we import outside things, we use absolute imports.

**Review by MrFlashAccount — APPROVED** [review]
LGTM aside a few moments above

File: app/dashboard/src/pages/dashboard/Settings/MembersTable.tsx (hunk 3 above)
Comment (somebody1234) [reply app/dashboard/src/pages/dashboard/Settings/MembersTable.tsx]: makes sense, i've moved `Settings.tsx` -> `Settings/index.tsx` and `Drive.tsx` -> `Drive/index.tsx` as well

**Comment by Frizi** [comment]
>     When navigating, the URL search parameters are lost (and so they are missing by the time GUI reads them).
>      -  GUI team should review whether this is an issue for the time being
>      -  We should probably rethink how CLI parameters should be passed to GUI (if at all)

I agree that we should modify this. GUI doesn't really have any need to read URL parameters directly now. Looks like all that we care about is carried through config provided by the dashboard. You can check all config usages by looking for `provideGuiConfig` and `injectGuiConfig`. There aren't many. We should simplify that whole config object to just simple props object, and remove all that complicated config reading logic. The dasboard would then fully own the URL through its router.

**Comment by somebody1234** [comment]
for posterity, the GUI config options are here:
https://github.com/enso-org/enso/blob/e609b1cff89cbaf0b2593601534fd55020ade066/app/gui2/src/config.json

going through each category:
- startup: passed manually by dashboard
- engine: passed manually by dashboard
- authentication: unused (auth is handled by dashboard)
- window:
  - `topBarOffset`: obsolete after the new redesign that switches from buttons to switch pages, to a tab bar. in the current develop, the GUI top bar and the tab bar are separate so there is no offset required to avoid overlap.
  - `vibrancy`: unused (should be handled by dashboard)

so i believe there should be no problem with omitting the URL search params then

<blockquote><div> GitHub</div><div><strong><a href="https://github.com/enso-org/enso/blob/e609b1cff89cbaf0b2593601534fd55020ade066/app/gui2/src/config.json">enso/app/gui2/src/config.json at e609b1cff89cbaf0b2593601534fd55020ade066 · enso-org/enso</a></strong></div><div>Hybrid visual and textual functional programming. Contribute to enso-org/enso development by creating an account on GitHub.</div></blockquote>

File: app/dashboard/src/layouts/CategorySwitcher/Category.ts [hunk 4]
```diff
@@ -1,34 +1,51 @@
...
+export type DriveCategory = (typeof DRIVE_CATEGORIES)[number]
 
-// This is REQUIRED, as `export default enum` is invalid syntax.
+// This is a function, even though it does not contain function syntax.
 // eslint-disable-next-line no-restricted-syntax
-export default Category
+export const isDriveCategory = includesPredicate(DRIVE_CATEGORIES)
+
+// ============================
+// === Globak configuration ===
```
Comment (indiv0) [inline app/dashboard/src/layouts/CategorySwitcher/Category.ts]: Globak -> Global

File: app/dashboard/src/layouts/CategorySwitcher.tsx [hunk 5]
```diff
@@ -172,16 +170,15 @@ export default function CategorySwitcher(props: CategorySwitcherProps) {
...
+  const navigate = useNavigate()
 
   const localBackend = backendProvider.useLocalBackend()
   /** The list of *visible* categories. */
   const categoryData = React.useMemo(
     () =>
       CATEGORY_DATA.filter((data) => {
         switch (data.category) {
-          case Category.local: {
+          case 'local': {
```
Comment (indiv0) [inline app/dashboard/src/layouts/CategorySwitcher.tsx:181]: StringlyTypeScript :P

File: app/dashboard/src/appUtils.tsx [hunk 6]
```diff
@@ -1,53 +1,64 @@
...
-  return SUBSCRIBE_PATH + '?plan=' + plan
+export function getUpgradeURL(plan: string): AppFullPath {
+  return `/subscribe?plan=${plan}`
 }
 
 /**
  * Build a Subscription URL for contacting sales.
  */
-export function getContactSalesURL(): string {
+export function getContactSalesURL(): `mailto:${string}` {
```
Comment (indiv0) [inline app/dashboard/src/appUtils.tsx:64]: ... you can use format strings as return types? TIL

File: app/dashboard/src/layouts/CategorySwitcher.tsx (hunk 5 above)
Comment (somebody1234) [reply app/dashboard/src/layouts/CategorySwitcher.tsx:181]: yeah. i guess the idea is mostly that enums are kinda unwieldy and don't offer any real benefit

File: app/dashboard/src/appUtils.tsx (hunk 6 above)
Comment (somebody1234) [reply app/dashboard/src/appUtils.tsx:64]: yeah, it's... somewhat (?) old at this point - they were introduced in TS 4.1 iirc:
https://www.typescriptlang.org/docs/handbook/2/template-literal-types.html

File: app/dashboard/src/appUtils.tsx [hunk 7]
```diff
@@ -1,53 +1,64 @@
...
+export function definePath(path: AppPath) {
+  return path
+}
+
 /** A {@link RegExp} matching all paths. */
 export const ALL_PATHS_REGEX = new RegExp(
-  `(?:${DASHBOARD_PATH}|${LOGIN_PATH}|${REGISTRATION_PATH}|${CONFIRM_REGISTRATION_PATH}|` +
-    `${FORGOT_PASSWORD_PATH}|${RESET_PASSWORD_PATH}|${SET_USERNAME_PATH}|${RESTORE_USER_PATH}|` +
-    `${SUBSCRIBE_PATH}|${SUBSCRIBE_SUCCESS_PATH})$`,
+  `/$|/(?:${APP_PATHS.map((path) => (path === 'editor' ? `${path}/` : path)).join('|')})\\b.*?$`,
```
Comment (indiv0) [inline app/dashboard/src/appUtils.tsx:44]: Isn't this more permissive since it matches things like `/registrationblah`, which wouldn't be accepted previously? Is this intentional?

Comment (somebody1234) [reply app/dashboard/src/appUtils.tsx:44]: shouldn't be because of the trailing `\b` - however it would technically accept trailing segments which i *think* is fine? its main function is as a heuristic to guess what path is the root path of the app

(although the `\b` *does* mean it does also match `/registration-blah` among other things...)

**Comment by PabloBuchu** [comment]
So something isn't right here. Tried to open local project and got.. Lemme know how time consuming fixing this might be.. is this PR very crucial for us right now?

Screenshot 2024-07-29 at 14 44 04

**Comment by somebody1234** [comment]
yeah it's not very crucial, will look into that issue though

File: app/dashboard/src/appUtils.tsx (hunk 7 above)
Comment (indiv0) [reply app/dashboard/src/appUtils.tsx:44]: image

Ah yeah I was testing with a `/` in the path.

**Comment by somebody1234** [comment]
worth noting though that that seems to be a WASM parser issue so i assume it shouldn't be caused by changes in this PR, is it consistently reproducible? (and consistently not reproducible on develop?)

couldn't repro in dev mode but i'm building the app rn to test the built appimage

**Comment by somebody1234** [comment]
huh nvm, i can repro in the built app, will investigate

**Comment by indiv0** [comment]
> worth noting though that that seems to be a WASM parser issue so i assume it shouldn't be caused by changes in this PR, is it consistently reproducible? (and consistently not reproducible on develop?)
> 
> couldn't repro in dev mode but i'm building the app rn to test the built appimage

Absolutely wild guess: something to do with the app path being matched incorrectly or modified incorrectly, then when the app tries to init the WASM bundle it makes a network request to the wrong location and thus can't fetch it and run it.

I'd check for failing network requests to `.wasm` paths or something and see if they're to different URLs than they were prior to this PR.

**Comment by PabloBuchu** [comment]
It works fine on develop so we probably missed some places with the old logic?

File: app/dashboard/src/utilities/appBaseUrl.ts [hunk 8]
```diff
@@ -1,5 +1,5 @@
 /** @file The base URL of the app. */
-import * as appUtils from '#/appUtils'
+import { ALL_PATHS_REGEX } from '#/appUtils'
 
 /** The base URL of the app. */
-export const APP_BASE_URL = location.pathname.replace(appUtils.ALL_PATHS_REGEX, '')
+export const APP_BASE_URL = location.pathname.replace(ALL_PATHS_REGEX, '')
```
Comment (indiv0) [inline app/dashboard/src/utilities/appBaseUrl.ts:5]: Is this stripping more than it should be or something so we have the wrong `APP_BASE_URL`?

File: app/dashboard/src/App.tsx [hunk 9]
```diff
@@ -122,14 +124,14 @@ LocalStorage.registerKey('inputBindings', {
...
+// === getRootUrl ===
+// ==================
 
-/** Returns the URL to the main page. This is the current URL, with the current route removed. */
-function getMainPageUrl() {
+/** Returns the URL to the app root. This is the current URL, with the current route removed. */
+function getRootUrl() {
   const mainPageUrl = new URL(window.location.href)
-  mainPageUrl.pathname = mainPageUrl.pathname.replace(appUtils.ALL_PATHS_REGEX, '')
+  mainPageUrl.pathname = appBaseUrl.APP_BASE_URL
```
Comment (indiv0) [inline app/dashboard/src/App.tsx:129]: Is this behaviour change intentional? It looks like `APP_BASE_URL` is now only computed once whereas before it would be computed every time this function was called?

File: app/dashboard/src/App.tsx [hunk 10]
```diff
@@ -356,7 +358,8 @@ function AppRouter(props: AppRouterProps) {
     }
   }, [localStorage, inputBindingsRaw])
 
-  const mainPageUrl = getMainPageUrl()
+  const mainPageUrl = getRootUrl()
+  mainPageUrl.pathname += '/drive'
```
Comment (indiv0) [inline app/dashboard/src/App.tsx]: Why do we need `/drive` here now when we didn't before?

File: app/dashboard/src/App.tsx (hunk 9 above)
Comment (somebody1234) [reply app/dashboard/src/App.tsx:129]: i think it's safe to assume the base URL at which the static app is hosted doesn't change over time :p

File: app/dashboard/src/utilities/appBaseUrl.ts (hunk 8 above)
Comment (somebody1234) [reply app/dashboard/src/utilities/appBaseUrl.ts:5]: i don't think so, the app seems to be working fine apart from the wasm stuff

**Comment by somebody1234** [comment]
> Absolutely wild guess: something to do with the app path being matched incorrectly or modified incorrectly

that doesn't seem to be it:

image

looks the wasm is being fetched correctly and instead it seems to be a race condition when fetching the wasm (the JS is trying to use the WASM before the module is ready). but... it's not like it's being fetched multiple times so not sure what might be going wrong here

File: app/dashboard/src/App.tsx (hunk 10 above)
Comment (somebody1234) [reply app/dashboard/src/App.tsx]: great question... i think it's no longer required because the Dashboard handles redirecting to `/drive` itself so it can safely be changed back

Comment (somebody1234) [reply app/dashboard/src/App.tsx]: umm, it seems like this was the issue...

**Comment by somebody1234** [comment]
seems like @indiv0 found the issue... @PabloBuchu should be ready for further testing now

File: app/dashboard/src/App.tsx (hunk 10 above)
Comment (somebody1234) [reply app/dashboard/src/App.tsx]: nvm, i was testing on dev mode which does work for some reason. will have to rebuild to see whether this helps any

File: app/dashboard/src/App.tsx (hunk 9 above)
Comment (indiv0) [reply app/dashboard/src/App.tsx:129]: ...fair

**Comment by indiv0** [comment]
> > Absolutely wild guess: something to do with the app path being matched incorrectly or modified incorrectly
> 
> that doesn't seem to be it:
> 
> image
> 
> looks the wasm is being fetched correctly and instead it seems to be a race condition when fetching the wasm (the JS is trying to use the WASM before the module is ready). but... it's not like it's being fetched multiple times so not sure what might be going wrong here

dumb question... is that WASM actually WASM? 605K seems small

**Comment by somebody1234** [comment]
yeah - the only thing in rust WASM now is the parser

**Comment by somebody1234** [comment]
never mind! it seems like it is indeed a GUI issue:
- https://github.com/enso-org/enso/pull/10688

**Comment by somebody1234** [comment]
@PabloBuchu merged latest develop so hopefully the issue will be gone now

**Review by indiv0 — APPROVED** [review]
LGTM (just review, no QA)

**Comment by somebody1234** [comment]
can confirm that the issue launching local projects is fixed after merge

**Comment by PabloBuchu** [comment]
1. This is how it behaves when you try to switch between opened project and local dashboard (ugly visible redirect). @somebody1234 this change seems much more unstable than we think. Do we really need to merge it before September release? 

https://github.com/user-attachments/assets/2869cdbe-6279-4013-9379-cb1e2bae46bb

**Comment by somebody1234** [comment]
closing as it's a (relatively minor) logic change with little to no benefit
